src/wp-includes/abilities-api/class-wp-ability.php

```diff
@@ -0,0 +1,588 @@
+<?php
+/**
+ * Abilities API
+ *
+ * Defines WP_Ability class.
+ *
+ * @package WordPress
+ * @subpackage Abilities API
+ * @since 6.9.0
+ */
+
+declare( strict_types = 1 );
+
+/**
+ * Encapsulates the properties and methods related to a specific ability in the registry.
+ *
+ * @since 6.9.0
+ *
+ * @see WP_Abilities_Registry
+ */
+class WP_Ability {
+
+	/**
+	 * The default value for the `show_in_rest` meta.
+	 *
+	 * @since 6.9.0
+	 * @var bool
+	 */
+	protected const DEFAULT_SHOW_IN_REST = false;
+
+	/**
+	 * The default ability annotations.
+	 * They are not guaranteed to provide a faithful description of ability behavior.
+	 *
+	 * @since 6.9.0
+	 * @var array<string,(bool|string)>
+	 */
+	protected static $default_annotations = array(
+		// Instructions on how to use the ability.
+		'instructions' => '',
+		// If true, the ability does not modify its environment.
+		'readonly'     => false,
+		/*
+		 * If true, the ability may perform destructive updates to its environment.
+		 * If false, the ability performs only additive updates.
+		 */
+		'destructive'  => true,
+		/*
+		 * If true, calling the ability repeatedly with the same arguments will have no additional effect
+		 * on its environment.
+		 */
+		'idempotent'   => false,
+	);
+
+	/**
+	 * The name of the ability, with its namespace.
+	 * Example: `my-plugin/my-ability`.
+	 *
+	 * @since 6.9.0
+	 * @var string
+	 */
+	protected $name;
+
+	/**
+	 * The human-readable ability label.
+	 *
+	 * @since 6.9.0
+	 * @var string
+	 */
+	protected $label;
+
+	/**
+	 * The detailed ability description.
+	 *
+	 * @since 6.9.0
+	 * @var string
+	 */
+	protected $description;
+
+	/**
+	 * The optional ability input schema.
+	 *
+	 * @since 6.9.0
+	 * @var array<string,mixed>
+	 */
+	protected $input_schema = array();
+
+	/**
+	 * The optional ability output schema.
+	 *
+	 * @since 6.9.0
+	 * @var array<string,mixed>
+	 */
+	protected $output_schema = array();
+
+	/**
+	 * The ability execute callback.
+	 *
+	 * @since 6.9.0
+	 * @var callable( mixed $input= ): (mixed|\WP_Error)
+	 */
+	protected $execute_callback;
+
+	/**
+	 * The optional ability permission callback.
+	 *
+	 * @since 6.9.0
+	 * @var callable( mixed $input= ): (bool|\WP_Error)
+	 */
+	protected $permission_callback;
+
+	/**
+	 * The optional ability metadata.
+	 *
+	 * @since 6.9.0
+	 * @var array<string,mixed>
+	 */
+	protected $meta;
+
+	/**
+	 * The ability category (required).
+	 *
+	 * @since 6.9.0
+	 * @var string
+	 */
+	protected $category;
+
+	/**
+	 * Constructor.
+	 *
+	 * Do not use this constructor directly. Instead, use the `wp_register_ability()` function.
+	 *
+	 * @access private
+	 *
+	 * @since 6.9.0
+	 *
+	 * @see wp_register_ability()
+	 *
+	 * @param string              $name The name of the ability, with its namespace.
+	 * @param array<string,mixed> $args An associative array of arguments for the ability. This should include:
```

Comment:
https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#1-1-parameters-that-are-arrays